text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,234 @@
+# Meta
+[meta]: #meta
+- Name: Offline Buildpackages
+- Start Date: 2020-04-27
+- Author(s): Dan Thornton
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+
+# Summary
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+We need to have a method to attach dependencies to buildpackages to support offline workflows. These dependencies should then be made available via the filesystem during the `build` phase.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+    - Remove need for two separate `ids` for online vs offline buildpacks
+    - Performance optimizations
+
+# What it is
+[what-it-is]: #what-it-is
+
+This provides a high level overview of the feature.
+
+- Define any new terminology.
+    - Here we add the concept of an offline buildpackage. This is just a buildpackage with additional dependencies added in layers.
+
+- Define the target persona:
+    - Buildpack user, Platform operator, Buildpack Author
+- Explaining the feature largely in terms of examples.
+
+If we want to support offline builds, with the current lifecycle & buildpackage spec we are forced to throw all dependencies inside of a buildpack archive. So an offline `ruby` buildpack would have a structure similar to the following:
+
+```
+cnb
+├─ buildpacks
+│   └─ ruby-buildpack
+│       └─ 1.2.3
+│           ├── bin
+│           │   ├── build
+│           │   └── detect
+│           ├── dependencies
+│           │   ├── ruby-v2.7.1
+│           │   └── ruby-v2.6.8
+│           └── buildpack.toml
+└── lifecycle
+```
+There are problems with the above:
+
+- Buildpacks packaged with dependencies & packaged without dependencies have different `sha256` values and therefore must have different buildpack `id` fields to avoid registry collision issues.
+
+We propose a `lifecycle` mechanism for exposing buildpack dependencies, and an extension to buildpackages that moves these dependencies out of buildpack archives.
+
+1) Buildpack dependencies should be made available during the `build` phase in a file tree rooted at `/cnb/deps`
+2) Buildpacks should be able to access dependencies they were packaged with via the `CNB_DEPS` environment variable.
+3) The [`package.toml`](https://buildpacks.io/docs/reference/package-config/) file should have additional metadata for each dependencies to specify
+    - the endpoint a dependency is exposed at via the lifecycle 
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Answering the following should outline the behavior of this feature.
+
+### How will dependencies appear on the filesystem during the `build` phase?
+
+Dependencies will be listed underneath `/cnb/deps/` with a schema that has the following properties:
+
+- dependency placement does not depend on the dependencies bits (no hashes, SHA256, ...).
```

Comment:
Can you add the reasoning explained during the WG today? /cc @sclevine @dwillist

Reply:
Believe this is clarified in the updated version [here](https://github.com/dwillist/rfcs/blob/offline-buildpackages/text/0000-offline-buildpackages.md#id--version-vs-hash-as-a-key)